blobstore/data/file_repo.go

```diff
@@ -0,0 +1,434 @@
+package data
+
+import (
+	"encoding/base64"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"strings"
+
+	"github.com/flynn/flynn/Godeps/_workspace/src/github.com/jackc/pgx"
+	"github.com/flynn/flynn/Godeps/_workspace/src/github.com/stevvooe/resumable"
+	"github.com/flynn/flynn/Godeps/_workspace/src/github.com/stevvooe/resumable/sha512"
+	"github.com/flynn/flynn/blobstore/backend"
+	"github.com/flynn/flynn/pkg/postgres"
+)
+
+const configEnvPrefix = "BACKEND_"
+
+func NewFileRepoFromEnv(db *postgres.DB) (*FileRepo, error) {
+	backends := []backend.Backend{backend.Postgres}
+	for _, env := range os.Environ() {
+		if !strings.HasPrefix(env, configEnvPrefix) {
+			continue
+		}
+		nameInfo := strings.SplitN(env, "=", 2)
+		name := strings.ToLower(strings.TrimPrefix(nameInfo[0], configEnvPrefix))
+		info := parseBackendInfo(nameInfo[1])
+		if info["backend"] != "s3" {
+			return nil, fmt.Errorf("blobstore: unknown backend %q for %s", info["backend"], name)
+		}
+		b, err := backend.NewS3(name, info)
+		if err != nil {
+			return nil, err
+		}
+		log.Printf("Configured additional backend: %s (%s)\n", name, info["backend"])
+		backends = append(backends, b)
+	}
+
+	defaultBackend := "postgres"
+	if d := os.Getenv("DEFAULT_BACKEND"); d != "" {
+		defaultBackend = d
+		var found bool
+		for _, b := range backends {
+			if b.Name() == d {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return nil, fmt.Errorf("error: unknown default backend %q", d)
+		}
+	}
+	return NewFileRepo(db, backends, defaultBackend), nil
+}
+
+func parseBackendInfo(s string) map[string]string {
+	info := make(map[string]string)
+	for _, token := range strings.Split(s, " ") {
+		kv := strings.SplitN(token, "=", 2)
+		info[kv[0]] = kv[1]
```

Comment:
this will panic with `index out of range` if the `s` is ill-formed (e.g `backend=s3 crash`)